app/gui/src/project-view/providers/interactionHandler.ts

```diff
@@ -44,37 +70,70 @@ export class InteractionHandler {
     if (this.isActive(interaction)) this.currentInteraction.value = undefined
   }
 
-  /** End the current interaction, if it is the specified instance. */
+  /**
+   * End the interaction, if it is currently active.
+   * Any children interactions of the given interaction will be ended as well.
+   */
   end(interaction: Interaction) {
-    if (this.isActive(interaction)) {
-      this.currentInteraction.value = undefined
-      interaction.end?.()
+    search(this.currentInteraction.value, interaction, (found, children) => {
+      this.currentInteraction.value = found.parentInteraction
+      found.end?.()
+      for (const interaction of children) {
+        interaction.end?.()
+      }
+    })
+  }
+
+  /** End all interactions, in the order from child to parent. */
+  endAll() {
+    let current = this.currentInteraction.value
+    while (current != null) {
+      current.end?.()
+      current = current.parentInteraction
     }
+    this.currentInteraction.value = undefined
   }
 
-  /** Cancel the current interaction, if it is the specified instance. */
+  /**
+   * Cancel the interaction, if it is currently active.
+   * Any children interactions of the given interaction will be cancelled as well.
+   */
   cancel(interaction: Interaction) {
-    if (this.isActive(interaction)) {
-      this.currentInteraction.value = undefined
-      interaction.cancel?.()
-    }
+    search(this.currentInteraction.value, interaction, (found, children) => {
+      this.currentInteraction.value = found.parentInteraction
+      found.cancel?.()
+      for (const interaction of children) {
+        interaction.cancel?.()
+      }
+    })
   }
 
-  /** TODO: Add docs */
-  handleCancel(): boolean {
+  /**
+   * Cancel all interactions, in the order from child to parent.
+   * @returns `true` if the current interaction was cancelled.
+   */
+  cancelAll(): boolean {
```

Comment:
Why in `cancelAll` we care about the canceling order, but in `cancel` we don't? Same question to `end`.

Reply:
We don’t care much now, but I think this property can be helpful in some situations. Made it more consistent.

Reply:
Only one thing: I imagine a situation, when someone in the future would like to change the order, but feel constrained, assuming that this guarantee is necessary. So I would keep the consistency, but remove the guarantee from docs.

Reply:
I think it is not an issue with our current workflow (we frequently refactor the code from the ground up, no matter the complications), but I agree that there is such a risk in principle. I have removed these comments.